### ISSUE TITLE
Getting results reported by paper on MBPP

### ISSUE THREAD

**Issue by lhaausing** [niansong1996/lever#5]
When we are implementing your model checkpoint given your hyperparameters (same as the yaml file) on MBPP, we get results as below:

Run summary:
binary_acc 0.72099
epoch 0
exec_acc_tmp_cat(100%) 0.61111
rerank_acc 0.61111
summing_agg_rerank_acc 0.63333
trainer/global_step 0
voting_agg_rerank_acc 0.61111

We are wondering how to reach the performance in the paper. Should we tweak hyperparameters, train our model, or even re-generate training/validation data from Codex?

It would be a huge help for our project if you could give up some tips. Thanks a lot for your help in advance.

**Comment by todpole3** [comment]
@lhaausing Sorry about the delay in response. Would you please clarify how the results above was obtained? Were you running the inference steps (https://github.com/niansong1996/lever#inference) using the released training data and could you share the training command you used?

**Comment by lhaausing** [comment]
Hi @todpole3 , thanks for responding.

We obtained the results mentioned above by using the checkpoint of MBPP(https://huggingface.co/niansong1996/lever-mbpp-codex) and used the released data (https://github.com/niansong1996/lever#data), specifically using the development set of MBPP. 
We did inference by running inference step:
`python finetuning/trainer.py validate --config finetuning/training_configs/verification/mbpp_verification.yaml --trainer.accelerator gpu --trainer.gpus 1`
We followed the MBPP config file (https://github.com/niansong1996/lever/blob/main/finetuning/training_configs/verification/mbpp_verification.yaml) and only changed `trainer.gpus `to `1` and `trainer.logger.init_args.entity/project` to my wandb account and project.

**Update**: 

We later also set `model.init_args.use_norm_gen_prob` to be `true` (as it was also mentioned in Table 7 of the paper). Here's the config after we changed to use normalized generation probability:
```
seed_everything: 333
trainer:
  gpus: 1
  gradient_clip_val: 1.0
  default_root_dir: &exp_name results/debug-tmp
  # val_check_interval: 0.5
  check_val_every_n_epoch: 1
  max_steps: &max_steps 10000
  # progress_bar_refresh_rate: 1
  num_sanity_val_steps: 0
  log_every_n_steps: 1
  logger+:
    - class_path: finetuning.lightning_modules.patches.patched_loggers.PatchedWandbLogger
      init_args:
        entity: lhaausing
        project: lever-reproduce-mbpp
        save_dir: *exp_name
        name: *exp_name
        log_model: False
        save_code: True
        offline: False
  callbacks+:
    - class_path: pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint
      init_args:
        monitor: rerank_acc
        mode: max
        filename: '{step}-{rerank_acc:.4f}-{binary_acc:.4f}'
        save_top_k: 3
        save_last: True
    - class_path: pytorch_lightning.callbacks.LearningRateMonitor
      init_args:
        logging_interval: step
    - class_path: pytorch_lightning.callbacks.progress.TQDMProgressBar
      init_args:
        refresh_rate: 1
        
  accelerator: gpu
  # strategy: deepspeed_stage_2
  strategy: ddp_find_unused_parameters_false
  # precision: 16
  accumulate_grad_batches: 2

model:
  class_path: lightning_modules.models.seq2seq_verification_model.EndVerificationModel
  init_args:
    transformer_model_name: &transformer niansong1996/lever-mbpp-codex
    executor_cls: execution.executors.MBPPExecutor
    max_gen_len: 16
    sampling_temp: 0.001
    # loss choices
    softmax_loss: false
    avg_loss_per_example: true
    contrastive: true
    mml: false
    eval_exec_result_agg: true
    exec_result_agg: false
    exec_result_agg_type: voting
    exec_result_agg_threshold: 0.0000001
    # filtering_top_k: 5
    # filtering_prob_threshold: 0.2
    gen_prob_coef: 1.0
    # max_generation_batches: 10
    max_batch_size: 4
    # eval_max_batch_size: 20
    gradient_ckpt: true
    # eval_greedy_search: true
    use_norm_gen_prob: true
    optimizer:
      init_args: 
        lr: 5.0e-5
        # lr: 0.0
        betas: 
          - 0.9
          - 0.999
        eps: 1.0e-8
        weight_decay: 0.1
    lr_scheduler:
      name: linear
      init_args:
        num_warmup_steps: 100
        num_training_steps: *max_steps

data:
  class_path: lightning_modules.datasets.mbpp_reader.MBPPEndVerificationDataModule
  init_args:
    transformer_model_name: *transformer
    batch_size: 1
    val_batch_size: 2
    # train_max_instances: 20
    # val_max_instances: 10
    train_set_init_args:
      file_path: data/mbpp/mbpp_codex_verification_train.jsonl
      # mask_context_loss: true
    val_set_init_args:
      file_path: data/mbpp/mbpp_codex_verification_dev.jsonl
    set_common_init_args:
      include_code: true
      include_exec_result: true
      ignore_no_negative_example: false
      ignore_no_positive_example: false
      use_exec_type_only: false
      use_exec_return_only: true
      use_final_exec_result_only: true
      # use_exec_passing_only: true
      state_str_func: state_simple_str_func
```
We finally obtained the results below:
```
binary_acc            0.72099369764328
exec_acc_tmp_cat(100%)     0.6555555462837219
rerank_acc           0.6555555462837219
summing_agg_rerank_acc     0.7555555701255798
voting_agg_rerank_acc     0.7222222089767456
```
The `summing_agg_rerank_acc` is close to what Table 5 reported (75.4%), but the `binary_acc` is still far from what is reported in Table 9 (77.3%).


We also fine-tuned a new verifier from `T5-large`, using the same hyperparameters (the one after setting `model.init_args.use_norm_gen_prob` to be `true`), but we got similar numbers for all evaluation metrics used above.

Thanks for your help in advance!

**Comment by niansong1996** [comment]
Hi @lhaausing, sorry for the late reply. To answer your questions-

So for the outputted metrics, `rerank_acc` is the **execution accuracy** after reranking w/o any aggregation (i.e., voting or summing), and `summing_agg_rerank_acc` should be the final performance we report in the paper (I think voting is not really used anywhere). 

As for `binary_acc`, sorry for the confusion but this simply refers to the binary classification accuracy of the verifier, and not the execution accuracy for the program that LEVER outputs in the end. It only denotes the accuracy of the verifier model itself, but LEVER uses both the probability of the verifier **and the generator** to select final candidates (hence `rerank_acc` and `summing_agg_rerank_acc` if aggregation is used).

For table 9, we are also reporting `summing_agg_rerank_acc` and not `binary_acc`. 

Hope this helps.

**Comment by lhaausing** [comment]
Thanks @niansong1996 ! That explains.

**Comment by lhaausing** [comment]
Hi @niansong1996 @todpole3 , thanks for the answer!

We can implement your code and get similar results as reported.

We only have one more question left, as asked in https://github.com/niansong1996/lever/issues/3 : how can we get file `../../../data/squall/db_info_wtq.json.` which is required for inference on wikitq?

Appreciate again!

**Comment by niansong1996** [comment]
Hi @lhaausing , I'm temporarily uploading the file here so you can use it. I will update the repo and instructions accordingly when I have more time to do so.

[db_info_wtq.json](https://github.com/niansong1996/lever/files/13631141/db_info_wtq.json)

**Comment by lhaausing** [comment]
Thanks @niansong1996 ! We will let you know if we have further questions.
